src/threadpool.c

```diff
@@ -29,350 +29,277 @@
-   * as the work queue.
-   */
-  if (pthread_atfork(NULL, NULL, &reset_once))
-    abort();
-#endif
-  init_threads();
-}
-
-
-void uv__work_submit(uv_loop_t* loop,
-                     struct uv__work* w,
-                     enum uv__work_kind kind,
-                     void (*work)(struct uv__work* w),
-                     void (*done)(struct uv__work* w, int status)) {
-  uv_once(&once, init_once);
-  w->loop = loop;
-  w->work = work;
-  w->done = done;
-  post(&w->wq, kind);
-}
+/* cleanup of the default_executor if necessary. */
+UV_DESTRUCTOR(static void cleanup()) {
+  unsigned int i;
 
+  if (!_fields.used)
+    return;
 
-static int uv__work_cancel(uv_loop_t* loop, uv_req_t* req, struct uv__work* w) {
-  int cancelled;
+  if (_fields.nworkers == 0)
+    return;
 
-  uv_mutex_lock(&mutex);
-  uv_mutex_lock(&w->loop->wq_mutex);
+  uv_mutex_lock(&_fields.mutex);
+  post(&_fields, &_fields.exit_message);
+  uv_mutex_unlock(&_fields.mutex);
 
-  cancelled = !QUEUE_EMPTY(&w->wq) && w->work != NULL;
-  if (cancelled)
-    QUEUE_REMOVE(&w->wq);
+  for (i = 0; i < _fields.nworkers; i++)
+    if (uv_thread_join(_fields.workers + i))
+      abort();
 
-  uv_mutex_unlock(&w->loop->wq_mutex);
-  uv_mutex_unlock(&mutex);
+  if (_fields.workers != _fields.default_workers)
+    uv__free(_fields.workers);
 
-  if (!cancelled)
-    return UV_EBUSY;
+  uv_mutex_destroy(&_fields.mutex);
+  uv_cond_destroy(&_fields.cond);
 
-  w->work = uv__cancelled;
-  uv_mutex_lock(&loop->wq_mutex);
-  QUEUE_INSERT_TAIL(&loop->wq, &w->wq);
-  uv_async_send(&loop->wq_async);
-  uv_mutex_unlock(&loop->wq_mutex);
-
-  return 0;
+  _fields.workers = NULL;
+  _fields.nworkers = 0;
 }
+#endif
 
+/******************************
+ * Default libuv threadpool, implemented using the executor API.
+*******************************/
 
-void uv__work_done(uv_async_t* handle) {
-  struct uv__work* w;
-  uv_loop_t* loop;
-  QUEUE* q;
-  QUEUE wq;
-  int err;
+static void uv__default_executor_submit(uv_executor_t* executor,
+                                        uv_work_t* req,
+                                        const uv_work_options_t* opts) {
+  struct default_executor_fields* fields;
+  struct uv__work* wreq;
 
-  loop = container_of(handle, uv_loop_t, wq_async);
-  uv_mutex_lock(&loop->wq_mutex);
-  QUEUE_MOVE(&loop->wq, &wq);
-  uv_mutex_unlock(&loop->wq_mutex);
+  assert(executor == &default_executor);
+  /* Make sure we are initialized internally. */
+  uv_once(&start_workers_once, start_workers);
 
-  while (!QUEUE_EMPTY(&wq)) {
-    q = QUEUE_HEAD(&wq);
-    QUEUE_REMOVE(q);
+  fields = (struct default_executor_fields *) executor->data;
+	assert(fields);
 
-    w = container_of(q, struct uv__work, wq);
-    err = (w->work == uv__cancelled) ? UV_ECANCELED : 0;
-    w->done(w, err);
-  }
-}
+  /* Put executor-specific data into req->reserved[0]. */
+  wreq = &req->work_req;
+  req->reserved[0] = wreq;
+  wreq->work = 0xdeadbeef; /* Non-NULL: "Not yet completed". */
```

Comment:
This isn't legal C if work is a function pointer.  Didn't the compiler warn?

Reply:
No comment.